### PR TITLE
[icubmod] dynamic_cast -> reinterpret_cast in Cfw2CanMessage::operator=

### PR DESCRIPTION
**Pull request by traversaro**
Rationale: `dynamic_cast` has a runtime overhead that we pay for each copied Cfw2CanMessage (rough estimate ~1% of CPU in pc104 use of robotInterface). 
Nobody is currently testing the success of this dynamic_cast by checking the `std::bad_cast` exception, so if it will ever fail both the dynamic_cast and the static_cast will cause a crash (however, I don't think this is possible with the current source base). 

I already tested this on the real robot, but given that this is a critical piece of code is better to review this change appropriately. 
cc @lornat75 @randaz81 @francesco-romano

### CODE REVIEW

**Comment by pattacini** [comment]
Go ahead then :)

**Comment by francesco-romano** [comment]
I add @drdanz for the review.

I am not completely confident about this patch.
Considering the implementation: `dynamic_cast` is the correct operator to be called because we are traversing the hierarchy down and not up: 

`static_cast < new_type > ( expression )`
>If new_type is a pointer or reference to some class D and the type of expression is a pointer or reference to its non-virtual base B, static_cast performs a downcast. 

Anyway in the original implementation i would switch to pointers instead of reference:
> Otherwise, the runtime check fails. If the dynamic_cast is used on pointers, the null pointer value of type new_type is returned. If it was used on references, the exception std::bad_cast is thrown.

There is one point which is not clear to me in the documentation: the word **non-virtual** base class in `static_cast` and the following point in the description of `dynamic_cast`:

>A downcast can also be performed with static_cast, which avoids the cost of the runtime check, but it's only safe if the program can guarantee (through some other logic) that the object pointed to by expression is definitely Derived.
Which does not talk about **non-virtual** inheritance.

As a last comment: I think the real problem resides in how the `operator=` is defined (and in its base class). I think in the current implementation something like the following code gives a wrong result:
```cpp
Cfw2CanMessage message1;
Cfw2CanMessage message2;

message1 = message2;

```

**Comment by traversaro** [comment]
Just to point out: the current use of this operator is guaranteeing that "that the object pointed to by expression is definitely Derived" (otherwise we would get a lot of `std::bad_cast`), and that's the rational behind the use of `static_cast`.

**Comment by francesco-romano** [comment]
The point for me unclear is the “non-virtual” derivation in the documentation.
In our case we have definitely a virtual polymorphism.

**Comment by traversaro** [comment]
Ah... understood. Feel free to reject the PR then.

**Comment by drdanz** [comment]
AFAIK A ``dynamic_cast`` on a reference cannot be checked, but you should enclose it in a try catch block instead... I would not recommend to use ``static_cast`` or ``reinterpret_cast``.

My approach would be to implement these 2:

```c++
virtual Cfw2CanMessage& operator=(const Cfw2CanMessage &l);
virtual Cfw2CanMessage& operator=(const CanMessage &l);
```

The 2nd one will overload the pure virtual assignment operator in the base class, and should just perform the ``dynamic_cast`` and call the other (or perform the copy if you are worried that the function call will slow down the execution).

This might save some ``dynamic_cast``s, but it depends on how the assignment operator is called, i.e. the ``CanMessage`` is called if you have a reference to a ``CanMessage`` or to some other class derived from ``CanMessage``, the other one is called **only** if you have a reference to a ``Cfw2CanMessage``. This unfortunately means that if you have a pointer to a ``CanMessage`` and you de-reference it, you will still get the ``CanMessage`` version, even if your object is actually a ``Cfw2CanMessage``.

**Comment by drdanz** [comment]
Also in assignment operator, usually the first thing you do is to check for self assignment for 2 reasons:

1. For efficiency.
2. Because a deep copy might cause problems if dynamic memory is used.

I don't know we have any self assignment, but I'd add a check here anyway.

**Comment by drdanz** [comment]
I found a relative thread:
http://www.cplusplus.com/forum/beginner/22130/

> You shouldn't have a virtual assignment operator.
>
> Consider the following:
> ```c++
> // assume childA and childB are derived from parent
> //   and parent has a virtual assignment operator
> 
> parent* a = new childA;
> parent* b = new childB;
> 
> *b = *a;  // !!  explode! 
>```
>
> With a virtual assignment operator, that will compile OK, but the consequences will be dire.
>
> An assignment should only progress if you know for a fact the two types are the same. And if you know for a fact the two types are the same, then you don't need polymorphism. 

But also:

>That's no reason to not declare it virtual. As long as derived classes only overload, not override, there shouldn't be a problem. What I'm trying to say is, there's no reason to declare operator=() as virtual, but i That's no reason to not declare it virtual. As long as derived classes only overload, not override, there shouldn't be a problem. What I'm trying to say is, there's no reason to declare operator=() as virtual, but it, in itself, doesn't hurt.t, in itself, doesn't hurt.

I don't know which reply is correct (if any), but I start to think that a pure virtual assignment operator is a bad idea.

I also found out that by defining it pure virtual, 

```c++
class A
{
public:
    virtual A& operator=(const A& rhs) = 0;
};
class B : public A
{
public
    virtual A& operator=(const A& rhs) { [...]; return this; }
}
```

You might think that the ``B`` class is ok, but this will actually fail to link:

```c++
int main(int argc, char *argv[])
{
    B b1;
    B b2;
    b1 = b2;
}
```

This is because according to the C++ standard

> if the subobject is of class type, the copy assignment operator for the class is used (as if by explicit qualification; that is, ignoring any possible virtual overriding functions in more derived classes); 

The default copy operator in this case takes precedence over the one defined because it is more explicit. Its first operation is to copy the base class calling the copy operator of the **base class** that is **pure virtual** (this for some reason that I ignore causes a linking error and not a build error as I would expect).

I would seriously consider to get rid of the pure virtual assignment operator..

**Comment by drdanz** [comment]
Very interesting reading about operator overloading:
http://arne-mertz.de/2015/01/operator-overloading-the-basics/
http://arne-mertz.de/2015/01/operator-overloading-common-practice/ (posted yesterday!)

**Comment by francesco-romano** [comment]
I agree with you @drdanz.
Because this would change the yarp interface I think we should hear also from @paulfitz

**Comment by paulfitz** [comment]
@francesco-romano which yarp interface? Not quite following, thought this thread/PR was an icubmod thing?

**Comment by francesco-romano** [comment]
@paulfitz 
`yarp::dev::CanMessage` class (https://github.com/robotology/yarp/blob/master/src/libYARP_dev/include/yarp/dev/CanBusInterface.h)

**Comment by paulfitz** [comment]
Ah, ok, handing over to @lornat75 in that case.

**Comment by lornat75** [comment]
Interesting discussion, nice to see yet another C++ tricky behavior...

I agree that a better approach could have been to define a separate function (i.e. clone or something like that) to stay away from the assignment operator. Another solution could have been to delegate copy to the base class at the cost of some virtual calls, which may be less expensive than dynamic_cast.

Anyway all this is very related to the canbus interfaces so I would not touch the interfaces.

Concerning the use of dynamic_cast, it seems more appropriate to me bc it can catch a problem by raising an exception (although we do not catch it), but in this specific case this seems to be unlikely to happen. Overall I would accept the pull request...

File: src/libraries/icubmod/cfw2Can/Cfw2Can.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ bool Cfw2Can::canGetErrors(CanErrors &err)
 
 CanMessage &Cfw2CanMessage::operator=(const CanMessage &l)
 {
-    const Cfw2CanMessage &tmp=dynamic_cast<const Cfw2CanMessage &>(l);
```
Comment (lornat75) [inline src/libraries/icubmod/cfw2Can/Cfw2Can.cpp]: I would add a comment to remind us that reinterpret_cast is used for efficiency but may hide bugs if the wrong object is passed on the call to Cfw2CanMessage::operator=(const CanMessage &l)

Comment (traversaro) [inline src/libraries/icubmod/cfw2Can/Cfw2Can.cpp]: done.

**Comment by drdanz** [comment]
:-1:

> Anyway all this is very related to the canbus interfaces so I would not touch the interfaces.

The more I think about it, the more I believe that 

* forcing to overload the ``operator=`` is just wrong and evil, each class should take care of calling its parent copy operator and then copying it's own stuff. If you wanted to add some private members to CanMessage, by making it pure virtual these could never be copied. Moreover the vtable lookup for the virtual function will cause some avoidable cpu cycles.
* ``reinterpret_cast`` is just evil here. What happens if for some reason the child class is changed not to implement the parent class as the first derived class? ``reinterpret_cast`` will not return a pointer to one actual child, but to some part of that structure, and using ``memcpy(msg ...)`` will either make it crash (if you are lucky) or copy some partial or random data.
* This looks like a workaround that will hide possible bugs, and will make it very hard to debug, instead of fixing the real issue that is copying one object from a pointer to the parent class and expecting that the whole object is copied. Different classes cannot be copied, unless you define the correct copy operator. "``<parent> = <parent>`` does not mean ``<child> = <child>``, it means *copy the ``<parent>`` part of ``<child>``*"

So, again I suggest to get rid of the pure virtual assignment operator in the canbus interface, and to fix the actual issues here.

Also an initial check for self-assignment is probably a good idea here, since I don't know what ``memcpy(msg, tmp.msg, sizeof(CFWCAN_MSG));`` does if ``*tmp == this``, but for sure it doesn't sound safe, since memcpy man page clearly states that:
> The memory areas must not overlap.

Anyway, too late, it was merged while I was writing this...

**Pull request merged**